docs/hosting/configuration-options.md
# Configuration options

PixivFE can be configured using environment variables or loading from a config file.

The order of precedence is as follows (from highest to lowest):

1. Environment variables
2. Configuration file
3. Default values

## Configuration file

By default, PixivFE tries to load the configuration from a `config.yml` file in the current working directory, if it exists.
You can create this file and put any of the configuration options below in it.

Here is an example of a minimal configuration file:

```yaml
basic:
  token:
    - 123456_arstdhnei
```

For all available configuration options, see the [example config.yml](https://gitlab.com/pixivfe/PixivFE/-/blob/v3/deploy/config.yml) in the repository.

You can specify a different path for the configuration file using the `-config` option.
The supplied config file must be in YAML format. Example:

```bash
# Build PixivFE and execute
go build
./pixivfe -config deploy/config.yml

# You can do `./pixivfe -h` to get help
```

It is also possible to use an environment variable to specify the configuration file:

```bash
# Build PixivFE and execute
go build
PIXIVFE_CONFIGFILE=deploy/config.yml ./pixivfe
```

## Environment variables

Any configuration option can be set as an environment variable. If a configuration option was both set in configuration file and/or in an environment variable, PixivFE will prioritize using environment variables.

For example:

```bash
PIXIVFE_TOKEN=YOUR_PIXIV_COOKIE PIXIVFE_PORT=8282 ./pixivfe
```

<!-- TODO: the following should link to v2 once v3 is merged -->

For all available configuration options, check out the [.env.example](https://gitlab.com/pixivfe/PixivFE/-/blob/v3/deploy/.env.example) in the repository.

## Basic options

**These options must be nested under a `basic:` block in `config.yml`.**

### `PIXIVFE_CONFIGFILE`

| YAML name | Environment variable | Required | Default        | Options |
| --------- | -------------------- | -------- | -------------- | ------- |
| -         | `PIXIVFE_CONFIGFILE` | No       | `./config.yml` | -       |

Specifies path to a YAML configuration file. This environment variable takes priority over the `-config` command line flag.

### `PIXIVFE_HOST`

| YAML name | Environment variable | Required | Default     | Options                |
| --------- | -------------------- | -------- | ----------- | ---------------------- |
| `host`    | `PIXIVFE_HOST`       | No       | `localhost` | Hostname or IP address |

The hostname or IP address that PixivFE should listen on and accept incoming connections from.

Mutually exclusive with `PIXIVFE_UNIXSOCKET`.

!!!note
    If you're **not using a reverse proxy** or **running PixivFE inside Docker**, you should set `PIXIVFE_HOST=0.0.0.0`. This will allow PixivFE to accept connections from any IP address or hostname. If you don't set this, PixivFE will refuse direct connections from other machines or devices on your network.

### `PIXIVFE_PORT`

| YAML name | Environment variable | Required | Default | Options     |
| --------- | -------------------- | -------- | ------- | ----------- |
| `port`    | `PIXIVFE_PORT`       | No       | `8282`  | Port number |

The port number that PixivFE should listen on and accept incoming connections from.

Mutually exclusive with `PIXIVFE_UNIXSOCKET`.

### `PIXIVFE_UNIXSOCKET`

| YAML name    | Environment variable | Required | Default | Options                                                              |
| ------------ | -------------------- | -------- | ------- | -------------------------------------------------------------------- |
| `unixSocket` | `PIXIVFE_UNIXSOCKET` | No       | -       | [UNIX socket path](https://en.wikipedia.org/wiki/Unix_domain_socket) |

The UNIX socket path that PixivFE should use.

Mutually exclusive with `PIXIVFE_HOST`/`PIXIVFE_PORT`.

### `PIXIVFE_TOKEN`

| YAML name | Environment variable | Required | Default | Options                                   |
| --------- | -------------------- | -------- | ------- | ----------------------------------------- |
| `token`   | `PIXIVFE_TOKEN`      | Yes      | -       | Comma-separated tokens (multiple allowed) |

`PHPSESSID` cookie(s) for authentication with the pixiv API.

Multiple tokens can be specified.

Environment variable format:

```sh
PIXIVFE_TOKEN=token1,token2,token3
```

Config file format:

```yaml
token:
  - token1
  - token2
```

### `PIXIVFE_TZ`

| YAML name | Environment variable | Required | Default   | Options                                                                          |
| --------- | -------------------- | -------- | --------- | -------------------------------------------------------------------------------- |
| `tz`      | `PIXIVFE_TZ`         | No       | `Etc/Utc` | [tz database name](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones) |

The timezone setting used to properly handle pixiv API endpoints that work using local time. This affects features like the rankings page and comments.

Set this to match the timezone of the IP address you're using to make pixiv API requests.

## Content proxy servers

**These options must be nested under a `contentProxies:` block in `config.yml`.**

pixiv requires `Referer: https://www.pixiv.net/` in the HTTP request headers to fetch images from their servers.

If any of these individual settings is not set, the built-in proxy will be used for that specific service.

For setting up your own proxy server, refer to [hosting an image proxy server](image-proxy-server.md). Alternatively, see the [list of public image proxies](../public-image-proxies.md) if you prefer to use an existing deployment.

### `PIXIVFE_IMAGEPROXY`

| YAML name    | Environment variable | Required | Default          | Options |
| ------------ | -------------------- | -------- | ---------------- | ------- |
| `imageProxy` | `PIXIVFE_IMAGEPROXY` | No       | (built-in proxy) | URL     |

The URL of a server that acts as a reverse proxy for i.pximg.net.

### `PIXIVFE_STATICPROXY`

| YAML name     | Environment variable  | Required | Default          | Options |
| ------------- | --------------------- | -------- | ---------------- | ------- |
| `staticProxy` | `PIXIVFE_STATICPROXY` | No       | (built-in proxy) | URL     |

The URL of a server that acts as a reverse proxy for s.pximg.net.

### `PIXIVFE_UGOIRAPROXY`

| YAML name     | Environment variable  | Required | Default          | Options |
| ------------- | --------------------- | -------- | ---------------- | ------- |
| `ugoiraProxy` | `PIXIVFE_UGOIRAPROXY` | No       | (built-in proxy) | URL     |

The URL of a server that acts as a reverse proxy for t-hk.ugoira.com.

## Token management

**These options must be nested under a `tokenManager:` block in `config.yml`.**

PixivFE implements exponential backoff for token management to manage rate limiting.

The following environment variables control how PixivFE manages token timeouts when a token encounters repeated failures. The backoff time for a token starts at the base timeout and doubles with each failure, up to the maximum backoff time.

### `PIXIVFE_TOKEN_LOAD_BALANCING`

| YAML name            | Environment variable           | Required | Default       | Options                                        |
| -------------------- | ------------------------------ | -------- | ------------- | ---------------------------------------------- |
| `tokenLoadBalancing` | `PIXIVFE_TOKEN_LOAD_BALANCING` | No       | `round-robin` | `round-robin`, `random`, `least-recently-used` |

Specifies the method for selecting tokens when multiple tokens are provided in `PIXIVFE_TOKEN`.

- `round-robin`: Tokens are used in a circular order.
- `random`: A random token is selected for each request.
- `least-recently-used`: The token that hasn't been used for the longest time is selected.

This option is useful when you have multiple pixiv accounts and want to distribute the load across them.

### `PIXIVFE_TOKEN_MAX_RETRIES`

| YAML name         | Environment variable        | Required | Default | Options |
| ----------------- | --------------------------- | -------- | ------- | ------- |
| `tokenMaxRetries` | `PIXIVFE_TOKEN_MAX_RETRIES` | No       | `5`     | Integer |

Maximum retry attempts before marking a token as unavailable.

### `PIXIVFE_TOKEN_BASE_TIMEOUT`

| YAML name          | Environment variable         | Required | Default  | Options                                                  |
| ------------------ | ---------------------------- | -------- | -------- | -------------------------------------------------------- |
| `tokenBaseTimeout` | `PIXIVFE_TOKEN_BASE_TIMEOUT` | No       | `1000ms` | [`time.Duration`](https://pkg.go.dev/time#ParseDuration) |

Initial backoff duration on token failure.

### `PIXIVFE_TOKEN_MAX_BACKOFF_TIME`

| YAML name             | Environment variable             | Required | Default   | Options                                                  |
| --------------------- | -------------------------------- | -------- | --------- | -------------------------------------------------------- |
| `tokenMaxBackoffTime` | `PIXIVFE_TOKEN_MAX_BACKOFF_TIME` | No       | `32000ms` | [`time.Duration`](https://pkg.go.dev/time#ParseDuration) |

Maximum backoff duration during exponential retry.

## API response caching

**These options must be nested under a `cache:` block in `config.yml`.**

PixivFE implements a caching system for API responses to improve performance. The cache uses a [Least Recently Used (LRU) eviction policy](<https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_Recently_Used_(LRU)>).

Each cache entry is stored with an expiration time. When a cached item is accessed, its expiration time is checked. If the item has expired, it is treated as a cache miss, and a new request is made to the Pixiv API.

To ensure that responses are properly isolated between different users, the cache key for each item is generated based on both the URL of the request and the value of the user's `pixivfe-Token` cookie.

### `PIXIVFE_CACHE_ENABLED`

| YAML name      | Environment variable    | Required | Default | Options |
| -------------- | ----------------------- | -------- | ------- | ------- |
| `cacheEnabled` | `PIXIVFE_CACHE_ENABLED` | No       | `false` | Boolean |

Controls whether the caching system is enabled.

When disabled, all requests will be sent directly to the Pixiv API without caching.

Other caching configuration variables will have no effect if this variable is set to `false`.

### `PIXIVFE_CACHE_SIZE`

| YAML name   | Environment variable | Required | Default | Options |
| ----------- | -------------------- | -------- | ------- | ------- |
| `cacheSize` | `PIXIVFE_CACHE_SIZE` | No       | `100`   | Integer |

Specifies the maximum number of items that can be stored in the LRU cache.

This limits the memory usage of the cache.

When the cache reaches this size, the least recently used items will be evicted to make room for new entries.

### `PIXIVFE_CACHE_TTL`

| YAML name  | Environment variable | Required | Default | Options                                                  |
| ---------- | -------------------- | -------- | ------- | -------------------------------------------------------- |
| `cacheTTL` | `PIXIVFE_CACHE_TTL`  | No       | `60m`   | [`time.Duration`](https://pkg.go.dev/time#ParseDuration) |

Specifies the default Time To Live (TTL) for cached items.

This is the duration for which an item remains valid in the cache before it's considered stale and needs to be fetched again from the pixiv API.

The TTL is applied to most API responses and can safely be set to a high value. Dynamic content such as Discovery and Newest is never cached.

## HTTP caching

**These options must be nested under a `httpCache:` block in `config.yml`.**

These cache control settings affect [HTTP caching](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching) behavior and are separate from PixivFE's internal API response cache.

### `PIXIVFE_CACHE_CONTROL_MAX_AGE`

| YAML name            | Environment variable            | Required | Default | Options                                                  |
| -------------------- | ------------------------------- | -------- | ------- | -------------------------------------------------------- |
| `cacheControlMaxAge` | `PIXIVFE_CACHE_CONTROL_MAX_AGE` | No       | `30s`   | [`time.Duration`](https://pkg.go.dev/time#ParseDuration) |

Controls the `max-age` directive in the Cache-Control response header for artwork pages.

This determines how long browsers should cache the page before revalidating with the server.

### `PIXIVFE_CACHE_CONTROL_STALE_WHILE_REVALIDATE`

| YAML name                          | Environment variable                           | Required | Default | Options                                                  |
| ---------------------------------- | ---------------------------------------------- | -------- | ------- | -------------------------------------------------------- |
| `cacheControlStaleWhileRevalidate` | `PIXIVFE_CACHE_CONTROL_STALE_WHILE_REVALIDATE` | No       | `60s`   | [`time.Duration`](https://pkg.go.dev/time#ParseDuration) |

Controls the `stale-while-revalidate` directive in the Cache-Control response header for artwork pages.

This allows browsers to show stale content while fetching a fresh version in the background.

## Request parameters

**These options must be nested under a `request:` block in `config.yml`.**

### `PIXIVFE_ACCEPTLANGUAGE`

| YAML name        | Environment variable     | Required | Default          | Options                                                                                                   |
| ---------------- | ------------------------ | -------- | ---------------- | --------------------------------------------------------------------------------------------------------- |
| `acceptLanguage` | `PIXIVFE_ACCEPTLANGUAGE` | No       | `en-US,en;q=0.5` | [Accept-Language value](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#syntax) |

The value of the `Accept-Language` header used for requests to pixiv's API.

Change this to modify the response language.

## Response behavior

**These options must be nested under a `response:` block in `config.yml`.**

### `PIXIVFE_EARLY_HINTS_RESPONSES_ENABLED`

| YAML name                    | Environment variable                    | Required | Default | Options |
| ---------------------------- | --------------------------------------- | -------- | ------- | ------- |
| `earlyHintsResponsesEnabled` | `PIXIVFE_EARLY_HINTS_RESPONSES_ENABLED` | No       | `false` | Boolean |

Controls whether PixivFE's internal HTTP server returns [HTTP `103 Early Hints` informational responses](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103).

Used to return `Link` headers to the client for preloading images while the server is busy preparing the main HTML response.

!!! warning
    Some reverse proxies may not handle informational responses such as HTTP 103 correctly.

### `PIXIVFE_REQUESTLIMIT`

| YAML name      | Environment variable   | Required | Default | Options |
| -------------- | ---------------------- | -------- | ------- | ------- |
| `requestLimit` | `PIXIVFE_REQUESTLIMIT` | No       | -       | Integer |

Set a request limit for the internal HTTP server per 30 seconds.

Set to an integer to enable the built-in rate limiter, e.g., `PIXIVFE_REQUESTLIMIT=15`.

It's recommended to enable rate limiting in the reverse proxy in front of PixivFE rather than using this.

## Features

**These options must be nested under a `feature:` block in `config.yml`.**

### `PIXIVFE_POPULAR_SEARCH_ENABLED`

| YAML name              | Environment variable             | Required | Default | Options |
| ---------------------- | -------------------------------- | -------- | ------- | ------- |
| `popularSearchEnabled` | `PIXIVFE_POPULAR_SEARCH_ENABLED` | No       | `false` | Boolean |

Controls whether searching by popularity for a given tag is enabled.

!!! warning
    This feature requires several API calls for each search.

    API response caching via `PIXIVFE_CACHE_ENABLED=true` is recommended when this is enabled.


config/config.go
// Copyright 2023 - 2025, VnPower and the PixivFE contributors
// SPDX-License-Identifier: AGPL-3.0-only

package config

import (
	"context"
	"errors"
	"fmt"
	"log"
	"net/url"
	"os"
	"strings"
	"time"
	_ "time/tzdata" // Import the timezone database for when the system timezone database is not available

	// TODO: figure out how to properly implement urlx
	// the implementation in 5f8b659b49 causes config.go to segfault due to a nil pointer dereference when the PIXIVFE_IMAGEPROXY env var is not set
	// "github.com/goware/urlx"
	"codeberg.org/pixivfe/pixivfe/server/tokenmanager"
	"codeberg.org/pixivfe/pixivfe/server/utils"
	"github.com/sethvargo/go-envconfig"
	"gopkg.in/yaml.v3"
)

var (
	GlobalConfig   ServerConfig // GlobalConfig is a global variable that exposes serverConfig.
	revision       string       // revision stores the current version's revision information.
)

// Placeholder/fallback values.
const (
	unknownRevision string = "unknown"
	revisionFormat  string = "date-hash[+dirty]"
)

// Limiter.DetectionMethod values.
const (
	NoneDetectionMethod      string = ""
	LinkTokenDetectionMethod string = "linktoken"
	TurnstileDetectionMethod string = "turnstile"
)

// Configration defaults.
const (
	version                                 string        = "v3.0.1"
	defaultHost                             string        = "localhost"
	defaultPort                             string        = "8282"
	defaultTimeZone                         string        = "Etc/UTC"
	defaultImageProxyStaging                string        = BuiltInImageProxyPath
	defaultStaticProxyStaging               string        = BuiltInStaticProxyPath
	defaultUgoiraProxyStaging               string        = BuiltInUgoiraProxyPath
	defaultTokenLoadBalancing               string        = "round-robin"
	defaultTokenMaxRetries                  int           = 5
	defaultTokenBaseTimeout                 time.Duration = 1000 * time.Millisecond
	defaultTokenMaxBackoffTime              time.Duration = 32000 * time.Millisecond
	defaultCacheEnabled                     bool          = false
	defaultCacheSize                        int           = 100
	defaultCacheTTL                         time.Duration = 60 * time.Minute
	defaultCacheControlMaxAge               time.Duration = 30 * time.Second
	defaultCacheControlStaleWhileRevalidate time.Duration = 60 * time.Second
	defaultAcceptLanguage                   string        = "en-US,en;q=0.5"
	defaultEarlyHintsResponsesEnabled       bool          = false
	defaultPopularSearchEnabled             bool          = false
	defaultRepoURL                          string        = "https://codeberg.org/PixivFE/PixivFE"
	defaultResponseSaveLocation             string        = "/tmp/pixivfe/responses"
	defaultLogLevel                         string        = "info"
	defaultLogFormat                        string        = "console"
	defaultLimiterEnabled                   bool          = false
	DefaultLimiterFilterLocal               bool          = false
	DefaultLimiterIPv4Prefix                int           = 24 // /24 network
	DefaultLimiterIPv6Prefix                int           = 48 // /48 network
	defaultLimiterCheckHeaders              bool          = true
	defaultLimiterDetectionMethod           string        = ""
)

var defaultLogOutputs = []string{"stdout"}

type ServerConfig struct {
	Basic struct {
		Host         string `env:"PIXIVFE_HOST,overwrite" yaml:"host"`
		Port         string `env:"PIXIVFE_PORT,overwrite" yaml:"port"`
		UnixSocket   string `env:"PIXIVFE_UNIXSOCKET" yaml:"unixSocket"`
		TimeZone     string `env:"PIXIVFE_TZ,overwrite" yaml:"timeZone"`
		TimeLocation *time.Location
		Token        []string `env:"PIXIVFE_TOKEN" yaml:"token"`
	}

	ContentProxies struct {
		RawImage  string  `env:"PIXIVFE_IMAGEPROXY,overwrite" yaml:"imageProxy"`
		Image     url.URL // For i.pximg.net
		RawStatic string  `env:"PIXIVFE_STATICPROXY,overwrite" yaml:"staticProxy"`
		Static    url.URL // For s.pximg.net
		RawUgoira string  `env:"PIXIVFE_UGOIRAPROXY,overwrite" yaml:"ugoiraProxy"`
		Ugoira    url.URL // For ugoira.com
	}

	TokenManager struct {
		TokenManager   *tokenmanager.TokenManager
		LoadBalancing  string        `env:"PIXIVFE_TOKEN_LOAD_BALANCING,overwrite" yaml:"tokenLoadBalancing"`
		MaxRetries     int           `env:"PIXIVFE_TOKEN_MAX_RETRIES,overwrite" yaml:"tokenMaxRetries"`
		BaseTimeout    time.Duration `env:"PIXIVFE_TOKEN_BASE_TIMEOUT,overwrite" yaml:"tokenBaseTimeout"`
		MaxBackoffTime time.Duration `env:"PIXIVFE_TOKEN_MAX_BACKOFF_TIME,overwrite" yaml:"tokenMaxBackoffTime"`
	}

	Cache struct {
		Enabled bool          `env:"PIXIVFE_CACHE_ENABLED,overwrite" yaml:"cacheEnabled"`
		Size    int           `env:"PIXIVFE_CACHE_SIZE,overwrite" yaml:"cacheSize"`
		TTL     time.Duration `env:"PIXIVFE_CACHE_TTL,overwrite" yaml:"cacheTTL"`
	}

	HTTPCache struct {
		MaxAge               time.Duration `env:"PIXIVFE_CACHE_CONTROL_MAX_AGE,overwrite" yaml:"cacheControlMaxAge"`
		StaleWhileRevalidate time.Duration `env:"PIXIVFE_CACHE_CONTROL_STALE_WHILE_REVALIDATE,overwrite" yaml:"cacheControlStaleWhileRevalidate"`
	}

	Request struct {
		AcceptLanguage string `env:"PIXIVFE_ACCEPTLANGUAGE,overwrite" yaml:"acceptLanguage"`
	}

	Response struct {
		EarlyHintsResponsesEnabled bool   `env:"PIXIVFE_EARLY_HINTS_RESPONSES_ENABLED,overwrite" yaml:"earlyHintsResponsesEnabled"`
		RequestLimit               uint64 `env:"PIXIVFE_REQUESTLIMIT" yaml:"requestLimit"`
	}

	Feature struct {
		PopularSearchEnabled bool `env:"PIXIVFE_POPULAR_SEARCH_ENABLED,overwrite" yaml:"popularSearchEnabled"`
	}

	Instance struct {
		Version      string
		Revision     string
		RevisionDate string
		RevisionHash string
		IsDirty      bool
		StartingTime string
		RepoURL      string `env:"PIXIVFE_REPO_URL,overwrite" yaml:"repoUrl"`
	}

	Development struct {
		InDevelopment        bool   `env:"PIXIVFE_DEV" yaml:"inDevelopment"`
		ResponseSaveLocation string `env:"PIXIVFE_RESPONSE_SAVE_LOCATION,overwrite" yaml:"responseSaveLocation"`
	}

	Log struct {
		Level   string   `env:"PIXIVFE_LOG_LEVEL,overwrite" yaml:"logLevel"`
		Outputs []string `env:"PIXIVFE_LOG_OUTPUTS,overwrite" yaml:"logOutputs"`
		Format  string   `env:"PIXIVFE_LOG_FORMAT,overwrite" yaml:"logFormat"`
	}

	Limiter struct {
		Enabled            bool     `env:"PIXIVFE_LIMITER_ENABLED,overwrite" yaml:"enabled"`
		PassIPs            []string `env:"PIXIVFE_LIMITER_PASS_IPS,overwrite" yaml:"passList"`
		BlockIPs           []string `env:"PIXIVFE_LIMITER_BLOCK_IPS,overwrite" yaml:"blockList"`
		FilterLocal        bool     `env:"PIXIVFE_LIMITER_FILTER_LOCAL,overwrite" yaml:"filterLocal"`
		IPv4Prefix         int      `env:"PIXIVFE_LIMITER_IPV4_PREFIX,overwrite" yaml:"ipv4Prefix"`
		IPv6Prefix         int      `env:"PIXIVFE_LIMITER_IPV6_PREFIX,overwrite" yaml:"ipv6Prefix"`
		CheckHeaders       bool     `env:"PIXIVFE_LIMITER_CHECK_HEADERS,overwrite" yaml:"checkHeaders"`
		DetectionMethod    string   `env:"PIXIVFE_LIMITER_DETECTION_METHOD,overwrite" yaml:"detectionMethod"`
		PingHMAC           string   `env:"PIXIVFE_LIMITER_PING_HMAC" yaml:"pingHMAC"`
		TurnstileSitekey   string   `env:"PIXIVFE_LIMITER_TURNSTILE_SITEKEY" yaml:"turnstileSitekey"`
		TurnstileSecretKey string   `env:"PIXIVFE_LIMITER_TURNSTILE_SECRET_KEY" yaml:"turnstileSecretKey"`
	}
}

func (cfg *ServerConfig) GetToken() string {
	token := cfg.TokenManager.TokenManager.GetToken()
	if token == nil {
		log.Println("[WARNING] All tokens are timed out. Using the first available token.")

		return cfg.Basic.Token[0]
	}

	return token.Value
}

func (cfg *ServerConfig) GetP_AB() string {
	token := cfg.TokenManager.TokenManager.GetToken()
	if token == nil {
		log.Println("[WARNING] All tokens are timed out. Using the first available token.")

		return cfg.Basic.Token[0]
	}

	return token.P_AB
}

func (cfg *ServerConfig) LoadConfig() error {
	parseCommandLineArgs()

	// Override config file path with environment variable if set
	if envVar := os.Getenv("PIXIVFE_CONFIGFILE"); envVar != "" {
		configFilePath = envVar
	}

	cfg.setInstanceInfo()

	cfg.loadFromDefaults()

	if err := cfg.loadFromYAML(); err != nil {
		return fmt.Errorf("error loading YAML config: %w", err)
	}

	if err := cfg.loadFromEnv(); err != nil {
		return fmt.Errorf("error loading environment variables: %w", err)
	}

	if err := cfg.validateAndSet(); err != nil {
		return fmt.Errorf("configuration validation and setting failed: %w", err)
	}

	cfg.initComponents()

	cfg.printConfiguration()

	return nil
}

func (cfg *ServerConfig) setInstanceInfo() {
	cfg.Instance.Version = version
	cfg.Instance.Revision = revision
	cfg.Instance.RevisionDate, cfg.Instance.RevisionHash, cfg.Instance.IsDirty = parseRevision(revision)
	cfg.Instance.StartingTime = time.Now().UTC().Format("2006-01-02 15:04")

	log.Printf("PixivFE %s, revision %s\n", cfg.Instance.Version, cfg.Instance.Revision)

	if revision == "" {
		log.Printf("[WARNING] REVISION is not set. Continuing with unknown revision information.\n")
	} else if cfg.Instance.RevisionDate == unknownRevision {
		log.Printf("[WARNING] REVISION format is invalid: %s. Expected format '%s'. Continuing with full revision as hash.\n", revision, revisionFormat)
	}
}

func (cfg *ServerConfig) loadFromDefaults() {
	cfg.Basic.Host = defaultHost
	cfg.Basic.Port = defaultPort
	cfg.Basic.TimeZone = defaultTimeZone
	cfg.Instance.RepoURL = defaultRepoURL
	cfg.Request.AcceptLanguage = defaultAcceptLanguage
	cfg.ContentProxies.RawImage = defaultImageProxyStaging
	cfg.ContentProxies.RawStatic = defaultStaticProxyStaging
	cfg.ContentProxies.RawUgoira = defaultUgoiraProxyStaging
	cfg.TokenManager.LoadBalancing = defaultTokenLoadBalancing
	cfg.TokenManager.MaxRetries = defaultTokenMaxRetries
	cfg.TokenManager.BaseTimeout = defaultTokenBaseTimeout
	cfg.TokenManager.MaxBackoffTime = defaultTokenMaxBackoffTime
	cfg.Cache.Enabled = defaultCacheEnabled
	cfg.Cache.Size = defaultCacheSize
	cfg.Cache.TTL = defaultCacheTTL
	cfg.HTTPCache.MaxAge = defaultCacheControlMaxAge
	cfg.HTTPCache.StaleWhileRevalidate = defaultCacheControlStaleWhileRevalidate
	cfg.Response.EarlyHintsResponsesEnabled = defaultEarlyHintsResponsesEnabled
	cfg.Feature.PopularSearchEnabled = defaultPopularSearchEnabled
	cfg.Development.ResponseSaveLocation = defaultResponseSaveLocation
	cfg.Log.Level = defaultLogLevel
	cfg.Log.Outputs = defaultLogOutputs
	cfg.Log.Format = defaultLogFormat
	cfg.Limiter.Enabled = defaultLimiterEnabled
	cfg.Limiter.CheckHeaders = defaultLimiterCheckHeaders
	cfg.Limiter.DetectionMethod = defaultLimiterDetectionMethod
	cfg.Limiter.FilterLocal = DefaultLimiterFilterLocal
	cfg.Limiter.IPv4Prefix = DefaultLimiterIPv4Prefix
	cfg.Limiter.IPv6Prefix = DefaultLimiterIPv6Prefix
}

func (cfg *ServerConfig) loadFromYAML() error {
	if configFilePath == "" {
		return nil
	}

	_, err := os.Stat(configFilePath)
	if os.IsNotExist(err) {
		log.Printf("No configuration file found at %s, skipping YAML config\n", configFilePath)

		return nil
	}

	ymlConf, err := os.ReadFile(configFilePath)
	if err != nil {
		return fmt.Errorf("failed to read configuration file %s: %w", configFilePath, err)
	}

	if err := yaml.Unmarshal(ymlConf, cfg); err != nil {
		return fmt.Errorf("failed to parse YAML from %s: %w", configFilePath, err)
	}

	log.Printf("Successfully loaded configuration from %s\n", configFilePath)

	return nil
}

func (cfg *ServerConfig) loadFromEnv() error {
	return envconfig.Process(context.Background(), cfg)
}

func (cfg *ServerConfig) validateAndSet() error {
	// Handle listener configuration
	if cfg.Basic.UnixSocket != "" {
		if cfg.Basic.Host != "" || cfg.Basic.Port != "" {
			return fmt.Errorf("unix socket configured - cannot specify Host (%s) and Port (%s) simultaneously",
				cfg.Basic.Host, cfg.Basic.Port)
		}
	} else {
		// Set TCP defaults
		if cfg.Basic.Host == "" {
			cfg.Basic.Host = defaultHost
			log.Printf("[INFO] Binding to default host: %s\n", cfg.Basic.Host)
		}

		if cfg.Basic.Port == "" {
			cfg.Basic.Port = defaultPort
			log.Printf("[INFO] Using default port: %s\n", cfg.Basic.Port)
		}
	}

	// Validate TimeZone
	location, err := time.LoadLocation(cfg.Basic.TimeZone)
	if err != nil {
		return fmt.Errorf("invalid time zone '%s': %w", cfg.Basic.TimeZone, err)
	}

	cfg.Basic.TimeLocation = location

	// Check tokens
	if len(cfg.Basic.Token) == 0 {
		return errors.New("no token supplied. Please supply at least one token")
	}

	// Validate image proxy
	if err := validateProxy(&cfg.ContentProxies.RawImage, BuiltInImageProxyPath, "image"); err != nil {
		return err
	}

	if cfg.ContentProxies.RawImage == BuiltInImageProxyPath {
		cfg.ContentProxies.Image = url.URL{Path: BuiltInImageProxyPath}
	} else {
		parsedURL, _ := url.Parse(cfg.ContentProxies.RawImage)
		cfg.ContentProxies.Image = *parsedURL
	}

	// Validate static proxy
	if err := validateProxy(&cfg.ContentProxies.RawStatic, BuiltInStaticProxyPath, "static"); err != nil {
		return err
	}

	if cfg.ContentProxies.RawStatic == BuiltInStaticProxyPath {
		cfg.ContentProxies.Static = url.URL{Path: BuiltInStaticProxyPath}
	} else {
		parsedURL, _ := url.Parse(cfg.ContentProxies.RawStatic)
		cfg.ContentProxies.Static = *parsedURL
	}

	// Validate ugoira proxy
	if err := validateProxy(&cfg.ContentProxies.RawUgoira, BuiltInUgoiraProxyPath, "ugoira"); err != nil {
		return err
	}

	if cfg.ContentProxies.RawUgoira == BuiltInUgoiraProxyPath {
		cfg.ContentProxies.Ugoira = url.URL{Path: BuiltInUgoiraProxyPath}
	} else {
		parsedURL, _ := url.Parse(cfg.ContentProxies.RawUgoira)
		cfg.ContentProxies.Ugoira = *parsedURL
	}

	// Validate RepoURL
	repoURL, err := utils.ValidateURL(cfg.Instance.RepoURL, "Repo")
	if err != nil {
		return fmt.Errorf("invalid repo URL: %w", err)
	}

	cfg.Instance.RepoURL = repoURL.String()

	// Validate TokenLoadBalancing
	switch cfg.TokenManager.LoadBalancing {
	case "round-robin", "random", "least-recently-used":
		// valid
	default:
		return fmt.Errorf("invalid TokenLoadBalancing value: %s (must be 'round-robin', 'random', or 'least-recently-used')",
			cfg.TokenManager.LoadBalancing)
	}

	// Skip validating Limiter configuration if it's not enabled
	if !cfg.Limiter.Enabled {
		return nil
	}

	// Validate DetectionMethod
	switch cfg.Limiter.DetectionMethod {
	case NoneDetectionMethod, LinkTokenDetectionMethod, TurnstileDetectionMethod:
		// valid
	default:
		return fmt.Errorf(
			"invalid Limiter.DetectionMethod: '%s'. Must be one of %q, %q, or %q",
			cfg.Limiter.DetectionMethod,
			NoneDetectionMethod,
			LinkTokenDetectionMethod,
			TurnstileDetectionMethod,
		)
	}

	// PingHMAC is required if a detection method that uses it is enabled
	if cfg.Limiter.DetectionMethod == LinkTokenDetectionMethod || cfg.Limiter.DetectionMethod == TurnstileDetectionMethod {
		if cfg.Limiter.PingHMAC == "" {
			return errors.New(
				"Limiter.PingHMAC is required when Limiter.DetectionMethod is " +
					LinkTokenDetectionMethod + " or " + TurnstileDetectionMethod,
			)
		}

		if len(cfg.Limiter.PingHMAC) < 32 {
			return errors.New("Limiter.PingHMAC must be at least 32 characters long")
		}
	}

	// Turnstile specific configuration
	if cfg.Limiter.DetectionMethod == TurnstileDetectionMethod {
		if cfg.Limiter.TurnstileSitekey == "" {
			return errors.New(
				"Limiter.TurnstileSiteKey is required when Limiter.DetectionMethod is " +
					TurnstileDetectionMethod,
			)
		}

		if cfg.Limiter.TurnstileSecretKey == "" {
			return errors.New(
				"Limiter.TurnstileSecretKey is required when Limiter.DetectionMethod is " +
					TurnstileDetectionMethod,
			)
		}
	}

	if cfg.Limiter.IPv4Prefix < 0 || cfg.Limiter.IPv4Prefix > 32 {
		return fmt.Errorf("IPv4 prefix must be between 0 and 32, got %d", cfg.Limiter.IPv4Prefix)
	}

	if cfg.Limiter.IPv6Prefix < 0 || cfg.Limiter.IPv6Prefix > 128 {
		return fmt.Errorf("IPv6 prefix must be between 0 and 128, got %d", cfg.Limiter.IPv6Prefix)
	}

	return nil
}

func (cfg *ServerConfig) initComponents() {
	cfg.TokenManager.TokenManager = tokenmanager.NewTokenManager(cfg.Basic.Token, cfg.TokenManager.MaxRetries, cfg.TokenManager.BaseTimeout, cfg.TokenManager.MaxBackoffTime, cfg.TokenManager.LoadBalancing)
}

// parseRevision extracts RevisionDate, RevisionHash, and IsDirty status from the Revision string.
func parseRevision(revision string) (string, string, bool) {
	// Check if Revision is empty
	if revision == "" {
		return unknownRevision, unknownRevision, false
	}

	// Check if Revision is marked as dirty (has uncommitted changes)
	isDirty := strings.HasSuffix(revision, "+dirty")
	if isDirty {
		revision = strings.TrimSuffix(revision, "+dirty")
	}

	// Split Revision into two parts, RevisionDate and RevisionHash
	parts := strings.Split(revision, "-")
	if len(parts) == 2 {
		return parts[0], parts[1], isDirty
	}

	// Return unknown date, full string as hash if format doesn't match date-hash
	return unknownRevision, revision, isDirty
}

// validateProxy validates ContentProxies.
func validateProxy(rawURL *string, defaultPath string, proxyType string) error {
	if *rawURL == defaultPath {
		return nil
	}

	_, err := utils.ValidateURL(*rawURL, proxyType+" proxy server")
	if err != nil {
		return fmt.Errorf("invalid %s proxy URL: %w", proxyType, err)
	}

	return nil
}
